### ISSUE TITLE
Gradient of tf.reduce_max and tf.nn.max_pool don't agree with each other, and theano

### ISSUE THREAD

**Issue by ppwwyyxx** [tensorflow/tensorflow#4886]
```python
a = tf.placeholder(dtype=tf.float32, name='a', shape=[4])
b = tf.reduce_max(a)
b2 = tf.nn.max_pool(tf.reshape(a, [1, 2, 2, 1]),
        [1,2,2,1], [1,1,1,1], 'VALID')[0,0,0,0]
c = tf.gradients([b], [a])[0]
c2 = tf.gradients([b2], [a])[0]

with tf.Session() as sess:
    v = np.asarray([1, 2, 2, 2], dtype='float32')
    print sess.run(c, feed_dict={a:v})  # 0, 0.3, 0.3, 0.3
    print sess.run(c2, feed_dict={a:v})  # 0, 1, 0, 0

import theano.tensor as T
import theano.tensor.signal.downsample as D
a = T.fvector('a')
b = T.max(a)
b2 = T.reshape(a, [1,2,2])
b2 = D.max_pool_2d(b2, [2,2])[0,0,0]
c = T.grad(b, a)
c2 = T.grad(b2, a)
print c.eval({a: v})    # 0, 1, 1, 1
print c2.eval({a: v})    # 0, 1, 1, 1
```
Bug or feature? 😦 Although as subgradient they all seem to make sense, but is there a reason to justify the difference of reduce_max and max_pool?

I saw the [comment](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L71) for reduce_max and it seems like the first one is feature.

**Comment by jart** [comment]
@girving Another gradients question.

**Comment by infnorm** [comment]
@ebrevdo can you take a look on this since @girving is not available right now.

**Comment by aselle** [comment]
@ebrevdo can you take a look on this since @girving is not available right now.

**Comment by ebrevdo** [comment]
Is this on GPU or not?  Are you using cudnn?

**Comment by ebrevdo** [comment]
Looks like cudnn maxpool returns `[0, 1, 0, 0]`.  xq; is there a flag that controls this?

**Comment by zheng-xq** [comment]
I think both tf.reduce_max and tf.maxpool are correct in their own sense. But the Theano's answer is not. 

When a backprop comes back into a max operation, and if there are multiple winners, the credit can only be shared among them. There is no constraint on which one should get the credit, and by how much. But they shouldn't all get the full credit. 

Consider this:
a = max(x, x, x). The correct answer is d(a)/d(x)=1, but Theano's answer leads to d(a)/d(x) = 3,  following the chain rule. 

Both tf.reduce_max and tf.maxpool gives somewhat self-consistent answers. One equally distributed it among all windows, while the other gives it to the first one. Both arbitrary choices, but nevertheless correct in their own sense. 

Since we pick the maxpool implementation from Cudnn, it's unclear that we can get all implementations consistent across all different libraries. So my perspective is that each implementation has to be self-consistent in that the total weight of a backprop value sums to 1. It might be a bit over-constrained if we require all operations to make the same arbitrary choice.

**Comment by ppwwyyxx** [comment]
Makes sense to me. Thanks!
